### PR TITLE
Problem: no relocatable rocksdb binary in CI

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #444
Solution:
- build relocatable rocksdb binary using nix
- add six rocksdb binaries to CI release flow:
  - rocksdb-x86_64-Darwin[-testnet]
  - rocksdb-x86_64-Linux[-testnet]
  - rocksdb-arm64-Linux[-testnet]

test run: https://github.com/yihuang/cronos/releases/tag/v0.0.28-test

## support tarballs

- `cronosd-tarball`
- `cronosd-testnet-tarball`
- `cronosd-goleveldb-tarball`
- `cronosd-testnet-goleveldb-tarball`

And run on different platform/arch to get more binaries.

##  package structure:
```
$ nix build github:yihuang/cronos/bundle-rocksdb#cronosd-exe
$ find ./result/
./result/
./result/exe
./result/exe/cronosd
./result/bin
./result/bin/cronosd
./result/lib
./result/lib/libc-2.33.so
./result/lib/libbz2.so.1
./result/lib/libzstd.so.1.5.2
./result/lib/liblz4.so.1
./result/lib/libstdc++.so.6.0.28
./result/lib/librocksdb.so.6.26.0
./result/lib/libpthread.so.0
./result/lib/ld-linux-aarch64.so.1
./result/lib/libdl-2.33.so
./result/lib/libsnappy.so.1
./result/lib/libpthread-2.33.so
./result/lib/libdl.so.2
./result/lib/libm-2.33.so
./result/lib/libc.so.6
./result/lib/libzstd.so.1
./result/lib/libz.so.1.2.12
./result/lib/libbz2.so.1.0.6
./result/lib/liblz4.so.1.9.3
./result/lib/ld-2.33.so
./result/lib/libsnappy.so.1.1.9
./result/lib/librocksdb.so.6
./result/lib/libm.so.6
./result/lib/libz.so.1
./result/lib/libgcc_s.so.1
./result/lib/libstdc++.so.6
```

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Review by JayT106 — APPROVED** [review]

**Review by adu-crypto — APPROVED** [review]

File: flake.nix [hunk 1]
```diff
@@ -44,5 +50,45 @@
...
+            "${final.gnutar}/bin/tar" cfzhv $out -C ${bundle} .
+          '';
+      } // (with final;
+        let
+          matrix = lib.cartesianProductOfSets {
+            db_backend = [ "goleveldb" "rocksdb" ];
+            network = [ "mainnet" "testnet" ];
+            relocatable = [ true false ];
+          };
+          binaries = builtins.listToAttrs (builtins.map
```
Comment (tomtau) [inline flake.nix:72]: does relocatable matter for goleveldb? I assume it's primarily for RocksDB use cases?
in any case, it's lazy, so not an issue if it's not used

File: flake.lock [hunk 2]
```diff
@@ -15,22 +15,94 @@
...
+        "type": "github"
+      },
+      "original": {
+        "owner": "NixOS",
+        "ref": "nixos-unstable",
+        "repo": "nixpkgs",
+        "type": "github"
+      }
+    },
+    "nixpkgs_2": {
```
Comment (tomtau) [inline flake.lock:69]: why are two sets of nixpkgs needed?

Comment (yihuang) [reply flake.lock:69]: I guess need to override nixpkgs in gomod2nix

File: flake.nix (hunk 1 above)
Comment (yihuang) [reply flake.nix:72]: In theory goleveldb version has libc dependency at runtime.
I was thinking if we can unify the process with nix, at least for Linux and Mac.

File: flake.lock (hunk 2 above)
Comment (yihuang) [reply flake.lock]: fixed

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -17,34 +17,69 @@ jobs:
...
           echo 'GITHUB_TOKEN=${{secrets.GITHUB_TOKEN}}' > .release-env
-      - name: draft release
+      - name: build and upload goreleaser binaries
         run: make release
+
+  rocksdb-x86_64:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+    needs: ["goreleaser"]
```
Comment (yihuang) [inline .github/workflows/release.yml:27]: I've tried to run goreleaser and nix building in parallel, but goreleaser has an issue to update existing draft release: https://github.com/goreleaser/goreleaser/issues/3072

**Comment by yihuang** [comment]
I've added the following binaries into CI release workflow:
- `rocksdb-x86_64-Darwin[-testnet]`
- `rocksdb-x86_64-Linux[-testnet]`
- `rocksdb-arm64-Linux[-testnet]`

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -17,34 +17,69 @@ jobs:
...
+    runs-on: ubuntu-latest
+    environment: release
+    steps:
       - uses: cachix/install-nix-action@v16
+      - run: |
+         DEBIAN_FRONTEND=noninteractive
+         sudo apt-get update -q -y && sudo apt-get install -q -y qemu-system-aarch64 qemu-efi binfmt-support qemu-user-static
+         mkdir -p ~/.config/nix
+         sudo bash -c "echo system-features = aarch64-linux arm-linux >> /etc/nix/nix.conf"
```
Comment (yihuang) [inline .github/workflows/release.yml:64]: didn't fully understand how, but it seems to work.
it use some binfmt trick on linux to build arm64 binary on x86_64 machine without doing complete cross-compiling.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -17,34 +17,69 @@ jobs:
...
+      - uses: cachix/cachix-action@v10
+        with:
+          name: cronos
+          signingKey: "${{ secrets.CACHIX_SIGNING_KEY }}"
+      - name: build linux/mac x86_64 rocksdb binaries
+        run: |
+          PLATFORM="$(uname -s)_x86_64"
+          FLAKE="github:${GITHUB_REPOSITORY}/${GITHUB_REF_NAME}"
+          nix build -L ${FLAKE}#cronosd-tarball
+          cp result cronos_${GITHUB_REF_NAME}-rocksdb_${PLATFORM}.tar.gz
```
Comment (allthatjazzleo) [inline .github/workflows/release.yml:41]:
```suggestion
          cp result cronos_${GITHUB_REF_NAME:1}-rocksdb_${PLATFORM}.tar.gz
```
remove the first character `v` to align with current release naming

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -17,34 +17,69 @@ jobs:
...
+          nix build -L ${FLAKE}#cronosd-testnet-tarball
+          cp result cronos_${GITHUB_REF_NAME}-testnet-rocksdb_${PLATFORM}.tar.gz
```
Comment (allthatjazzleo) [inline .github/workflows/release.yml:43]:
```suggestion
          cp result cronos_${GITHUB_REF_NAME:1}-testnet-rocksdb_${PLATFORM}.tar.gz
```

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -17,34 +17,69 @@ jobs:
...
-          mkdir ./bin
-          nix bundle --bundler github:NixOS/bundlers .
-          mv ./arx ./bin/cronosd
-          tar cfzh "${MAINNET_TAR}" ./bin
-          nix bundle --bundler github:NixOS/bundlers .#cronosd-testnet
-          mv ./arx ./bin/cronosd
-          tar cfzh "${TESTNET_TAR}" ./bin
+          FLAKE="github:${GITHUB_REPOSITORY}/${GITHUB_REF_NAME}"
+          nix build -L --extra-platforms aarch64-linux ${FLAKE}#packages.aarch64-linux.cronosd-tarball
+          cp result cronos_${GITHUB_REF_NAME}-rocksdb_${PLATFORM}.tar.gz
```
Comment (allthatjazzleo) [inline .github/workflows/release.yml:75]:
```suggestion
          cp result cronos_${GITHUB_REF_NAME:1}-rocksdb_${PLATFORM}.tar.gz
```

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -17,34 +17,69 @@ jobs:
...
+          nix build -L --extra-platforms aarch64-linux ${FLAKE}#packages.aarch64-linux.cronosd-testnet-tarball
+          cp result cronos_${GITHUB_REF_NAME}-testnet-rocksdb_${PLATFORM}.tar.gz
```
Comment (allthatjazzleo) [inline .github/workflows/release.yml:77]:
```suggestion
          cp result cronos_${GITHUB_REF_NAME:1}-testnet-rocksdb_${PLATFORM}.tar.gz
```

**Pull request merged**